pyron-core/src/main/scala/com/cloudentity/pyron/rule/PreparedPathRewrite.scala

```diff
@@ -72,24 +74,27 @@ object PreparedPathRewrite {
 
   def applyRewrite(path: String, rewrite: PreparedPathRewrite): Option[AppliedPathRewrite] = for {
     regexMatch <- rewrite.regex.findFirstMatchIn(path)
-    targetPath = rewritePath(rewrite.rewritePattern, regexMatch)
-    pathParams = getPathParams(rewrite, regexMatch)
+    pathParams = getPathParams(rewrite.rewriteMap, regexMatch)
+    targetPath = rewritePath(pathParams, rewrite.rewritePattern)
   } yield AppliedPathRewrite(path, targetPath, pathParams, from = rewrite)
 
-  private[rule] def rewritePath(rewritePattern: String, regexMatch: Regex.Match): String =
-    (1 to regexMatch.groupCount).map(i => (i, regexMatch.group(i))).foldLeft(rewritePattern) {
-      case (rew, (idx, value)) => rew.replace(s"{$idx}", value)
+  private[rule] def rewritePath(pathParams: PathParams, rewritePattern: String): String = {
+    pathParams.value.foldLeft(rewritePattern) {
+      case (rew, (param, value)) => rew.replace(s"{$param}", value)
     }
+  }
 
-  private[rule] def getPathParams(rewrite: PreparedPathRewrite, regexMatch: Regex.Match): PathParams = {
-    val params = for {
-      i <- (2 - rewrite.groupCount) until rewrite.groupCount
-    } yield if (i > 0) {
-      rewrite.paramName(i).getOrElse(s"$i") -> regexMatch.group(i)
-    } else {
-      rewrite.paramName(i - 1).getOrElse(s"${i - 1}") -> regexMatch.group(rewrite.groupCount + i - 1)
-    }
-    PathParams(params.toMap)
+  private[rule] def getRewriteMap(rewritePattern: String, paramNames: List[(String, Int)], groupCount: Int): Map[String, Int] = {
```

Comment:
This method is not easy to understand. What does it try to achieve? Why do we iterate over `(2 - groupCount) until groupCount`? Can we implement it differently without using low-level operations on indexes?

Reply:
I agree it's not straightforward, it's a bit optimized to do everything in one pass. It could be done more verbose, I'm not sure if it would make it much simpler.

I will explain it here. 

Based on regex implementation, capture group count is number of capturing parentheses pairs + 1, because group 0 is the entire match. However, we are not really interested in group 0, containing the entire match.
For our purposes it is hardly useful and also difficult to handle, because it needs to be treated differently depending on weather we keep or drop the prefix. User can still capture entire path, explicitly with (.*) into group 1. Therefore, in my opinion it's best to ignore group 0. So for us the "usefulGroupCount' is groupCount - 1.

With this out of the way, say the "usefulGroupCount" is 3, we want to check if within rewrite pattern there are any of: {1}, {2}, {3} which reference capture groups counting left-to-right (from the start) but also {-1}, {-2}, {-3} which reference capture groups counting right-to-left (from the end). For positive numbers {1} maps to capture group 1, {2} to capture group 2, and {3} to capture group 3. For negative numbers {-1} maps to capture group 3, {-2} to capture group {2} and {-3} to capture group 1. As we see here, for negative refs, effective capture group depends on the total groups count, as {-1} will turn into "usefulGroupCount" and remaining refs will also change.

We add only the actually referenced in rewrites numeric mappings into the "numericRefs".
We add all (used or not in rewrites) named groups into the "namedParams", because they may still be used in plugins etc.
We add both of the above into the rewriteMap, and PathParams of each individual request are populated based on this mapping.

I think, instead of providing lengthy comment, the tests are maybe the best way to explain it. I added a couple of tests to demonstrate it. Otherwise we might comment the function with above explanation, but I think it's better to just explain through tests because comments tend to get out of sync quickly and become false info.